# Embeddings With Sentence-Transformers

We've worked through creating our embeddings using the `transformers` library - and at times it can be quite involved. Now, it's important to understand the steps, but we can make life easier by using the `sentence-transformers` library.

We'll work through the same process - but using `sentence-transformers` instead.

In [1]:
sentences = [
    "Three years later, the coffin was still full of Jello.",
    "The fish dreamed of escaping the fishbowl and into the toilet where he saw his friend go.",
    "The person box was packed with jelly many dozens of months later.",
    "Standing on one's head at job interviews forms a lasting impression.",
    "It took him a month to finish the meal.",
    "He found a leprechaun in his walnut shell."
]

# thanks to https://randomwordgenerator.com/sentence.php

Initialize our model:

In [4]:
!poetry add sentence-transformers

Using version ^2.2.2 for sentence-transformers

Updating dependencies
Resolving dependencies... (121.0s)Resolving dependencies... (67.9s)Resolving dependencies... (73.3s)Resolving dependencies... (79.0s)Resolving dependencies... (94.7s)Resolving dependencies... (106.9s)Resolving dependencies... (111.9s)Resolving dependencies... (117.2s)

  SolverProblemError

  Because no versions of sentence-transformers match >2.2.2,<3.0.0
   and sentence-transformers (2.2.2) depends on transformers (>=4.6.0,<5.0.0), sentence-transformers (>=2.2.2,<3.0.0) requires transformers (>=4.6.0,<5.0.0).
  And because farm (0.6.2) depends on transformers (4.1.1), sentence-transformers (>=2.2.2,<3.0.0) is incompatible with farm (0.6.2).
  And because farm-haystack (0.7.0) depends on farm (0.6.2)
   and no versions of farm-haystack match >0.7.0,<0.8.0, sentence-transformers (>=2.2.2,<3.0.0) is incompatible with farm-haystack (>=0.7.0,<0.8.0).
  So, because dev depends on both farm-haystack (^0.7.0) and sentence-

In [2]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('bert-base-nli-mean-tokens')

ModuleNotFoundError: No module named 'sentence_transformers'

Encode the sentences:

In [3]:
sentence_embeddings = model.encode(sentences)

In [4]:
sentence_embeddings.shape

(6, 768)

In [5]:
sentence_embeddings

array([[ 0.07464128,  0.8635731 ,  0.18004653, ...,  0.7732613 ,
         1.7246002 , -0.18037328],
       [-0.37150592,  0.97317815,  1.0839957 , ..., -0.25496134,
        -0.275705  ,  0.03565905],
       [-0.5025152 ,  0.7944973 , -0.12324826, ...,  0.14401469,
         0.97013956, -0.17904879],
       [-0.01328272,  0.9773259 ,  1.451448  , ..., -0.8458191 ,
        -1.4004992 , -0.41184014],
       [-0.20161639,  0.05962702,  0.86030865, ..., -0.00991616,
         0.8431719 , -0.08434674],
       [-0.21294485,  1.0176604 , -0.88306004, ...,  0.7370991 ,
         0.19437043, -0.30156055]], dtype=float32)

And no we have our sentence embeddings - a much quicker approach. We then compare just as we did before using cosine similarity:

In [6]:
from sklearn.metrics.pairwise import cosine_similarity

Let's calculate cosine similarity for sentence `0`:

In [7]:
cosine_similarity(
    [sentence_embeddings[0]],
    sentence_embeddings[1:]
)

array([[0.33088642, 0.7218851 , 0.17474297, 0.44715378, 0.55473834]],
      dtype=float32)

These similarities translate to almost the exact same values as we calculated before:

| Index | Sentence | Similarity (before) | New similarity |
| --- | --- | --- | --- |
| 1 | "The fish dreamed of escaping the fishbowl and into the toilet where he saw his friend go." | 0.3309 | 0.3309 |
| 2 | "The person box was packed with jelly many dozens of months later." | 0.7219 | 0.7219 |
| 3 | "Standing on one's head at job interviews forms a lasting impression." | 0.1748 | 0.174**7** |
| 4 | "It took him a month to finish the meal." | 0.4471 | 0.447**2** |
| 5 | "He found a leprechaun in his walnut shell." | 0.5548 | 0.554**7** |

So, using `sentence-transformers` can make life much easier. But either option produces the same outcome.